hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/transfer/AssociateTokenRecipientsStep.java

```diff
@@ -88,10 +91,34 @@ public void doIn(@NonNull final TransferContext transferContext) {
                 final var nft = nftStore.get(tokenId, aa.serialNumber());
                 validateTrue(nft != null, INVALID_NFT_ID);
 
-                if (nft.hasOwnerId()) {
-                    validateTrue(nft.ownerId().equals(senderId), SENDER_DOES_NOT_OWN_NFT_SERIAL_NO);
-                } else {
-                    validateTrue(token.treasuryAccountId().equals(senderId), SENDER_DOES_NOT_OWN_NFT_SERIAL_NO);
+                // First check for an allowance or spender ID (different from the owner) for this NFT
+                final var nftOwnerId = nft.hasOwnerId() ? nft.ownerIdOrThrow() : null;
+                final var nftSpenderId = nft.hasSpenderId() ? nft.spenderIdOrThrow() : null;
+                boolean allowanceFound = false;
+                if (nftOwnerId != null && !senderId.equals(nftOwnerId)) {
+                    final var senderAcct =
+                            getIfUsable(senderId, accountStore, handleContext.expiryValidator(), INVALID_ACCOUNT_ID);
+                    final var approvedNftAllowances = senderAcct.approveForAllNftAllowancesOrElse(emptyList());
+                    final var nftAllowanceId = AccountApprovalForAllAllowance.newBuilder()
+                            .tokenId(tokenId)
+                            .spenderId(senderId)
+                            .build();
+                    if (approvedNftAllowances.contains(nftAllowanceId)) {
+                        allowanceFound = true;
+                    } else {
+                        validateTrue(senderId.equals(nftSpenderId), SPENDER_DOES_NOT_HAVE_ALLOWANCE);
+                        allowanceFound = true;
+                    }
+                }
+
+                // If no allowance or spender is found, check the NFT's owner
+                if (!allowanceFound) {
+                    if (nft.hasOwnerId()) {
+                        validateTrue(nft.ownerIdOrThrow().equals(senderId), SENDER_DOES_NOT_OWN_NFT_SERIAL_NO);
+                    } else {
+                        // If there's no owner, this NFT's only authorized sender is the token treasury
+                        validateTrue(token.treasuryAccountId().equals(senderId), SENDER_DOES_NOT_OWN_NFT_SERIAL_NO);
```

Comment:
I think this logic is already there in `NFTOwnersChangeStep`. Not sure if we need these here

Reply:
I refactored the code so the same check is being done in both places, without duplicating the logic. Good call! 👍